lib/puppet/parser/compiler.rb

```diff
@@ -328,8 +328,8 @@ def prune_node_catalog
   # @api private
   def evaluate_applications
     @applications.each do |app|
-      mapping = app.parameters[:nodes].value
       components = []
+      mapping = app.parameters[:nodes] ? app.parameters[:nodes].value : {}
```

Comment:
While at it, think we need to protect against a user doing `nodes => true` or `nodes => 'blah'` by checking that it is indeed a hash or nil. The code above would be fooled by a truthy value and trying to lookup :nodes. (Not sure if an application guards against that).